### PR TITLE
Develop/rmf demo panel: a react gui with flask server in place of current rmf panel

### PR DESCRIPTION
**Pull request by vallq**
What's new:
- React frontend with simplified form fields for different requests
- This works with the different demo worlds according to the `dashboard_config.json` files available 

https://user-images.githubusercontent.com/28668944/103737140-8bd97700-502c-11eb-9798-f553dfa2236e.mp4

@Yadunund @aaronchongth @tanyouliang95 comments and feedback would be appreciated!

### CODE REVIEW

**Comment by Yadunund** [comment]
@vallq could we target the merge to `develop/dispatcher_demo` branch?

File: rmf_demo_panel/task_list/clinic_tasks.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[
+{ "task_type":"Loop", "start_time":0, "description": {"num_loops":5, "start_name":"L1_left_negative_air_pressure", "finish_name":"L2_left_nurse_center"}},
+{ "task_type":"Loop", "start_time":1, "description": {"num_loops":5, "start_name":"L1_left_nurse_center", "finish_name":"L2_south_couter"}},
```
Comment (tanyouliang95) [inline rmf_demo_panel/task_list/clinic_tasks.json:3]: I made a typo here previously: should rename `L2_south_couter` as `L2_south_counter`

File: rmf_demo_panel/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# RMF Demo Panel
+
+This will be a replacement of rviz's [rmf_panel](https://github.com/osrf/rmf_schedule_visualizer)
+
+## Installation
+Dependencies
+ - rmf_core: `feature/task-dispatcher` branch
```
Comment (tanyouliang95) [inline rmf_demo_panel/README.md:7]: We just merged in `feature/task-dispatcher` into `develop/dispatcher_demo`. Would appreciate if you can also update this. Thanks

**Review by tanyouliang95 — CHANGES_REQUESTED** [review]
Thanks for the PR. Would also like to point out the setup instruction: [here](https://github.com/osrf/rmf_demos/tree/develop/rmf_demo_panel/rmf_demo_panel).

Tested the 3 world scenarios, all works fine. One small issue that I notice during my brief test, the top `Time is XXX` is not showing up on my web panel. Other than that, LGTM.

File: rmf_demo_panel/task_list/clinic_tasks.json (hunk 1 above)
Comment (vallq) [reply rmf_demo_panel/task_list/clinic_tasks.json:3]: ok will update!

File: rmf_demo_panel/README.md (hunk 2 above)
Comment (vallq) [reply rmf_demo_panel/README.md:7]: right on it.

just wanted to check w you, when you mean `Time is XXX` does not show up, it's only the numerals not showing up or the entire line including `Time is`??

**Comment by tanyouliang95** [comment]
The above-mentioned issue on Time visualization is working. User will need to ensure that the local flask server's `flask_socketio.__version__` is `5.X`

File: demos/launch/dispatcher.launch.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?xml version='1.0' ?>
```
Comment (Yadunund) [inline demos/launch/dispatcher.launch.xml:1]: I recommend copying the arguments and the code block under "Dispatcher GUI Server"  into `common.launch.xml` under this block https://github.com/osrf/rmf_demos/blob/7bce0433f67eae0719324a5fe8c63a905dce556b/demos/launch/common.launch.xml#L46-L52

We can then delete this file

File: rmf_demo_panel/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# RMF Demo Panel
+
+This will be a replacement of rviz's [rmf_panel](https://github.com/osrf/rmf_schedule_visualizer)
```
Comment (Yadunund) [inline rmf_demo_panel/README.md:3]: The [RMF Panel](https://github.com/osrf/rmf_demos/tree/develop/dispatcher_demo/rmf_rviz_plugin) is located in `rmf_demos` repo and not `rmf_schedule_visualizer`. I'd recommend just deleting this line since we have plans to delete the old package anyways.

File: rmf_demo_panel/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+## Installation
+Dependencies
+ - rmf_core: [`develop/dispatcher_demo`](https://github.com/osrf/rmf_demos/tree/develop/dispatcher_demo) branch
+
+Setup `rmf_demo_panel`
+```bash
```
Comment (Yadunund) [inline rmf_demo_panel/README.md:10]: Does the user require `nodejs` to install the panel? If so let's add instructions to install the same.

File: rmf_demo_panel/package.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>rmf_demo_panel</name>
+  <version>0.0.0</version>
```
Comment (Yadunund) [inline rmf_demo_panel/package.xml:5]:
```suggestion
  <version>1.0.0</version>
```

File: rmf_demo_panel/rmf_demo_panel/static/dashboard_config.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (Yadunund) [inline rmf_demo_panel/rmf_demo_panel/static/dashboard_config.json:1]: I see that there are `dashboard_config.json` files for `office`, `airport` and `clinic` worlds. What is the purpose of this file?

File: rmf_demo_panel/rmf_demo_panel/templates/index.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html lang="en" dir="ltr">
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+<script src="//cdnjs.cloudflare.com/ajax/libs/socket.io/2.2.0/socket.io.js"></script>
```
Comment (Yadunund) [inline rmf_demo_panel/rmf_demo_panel/templates/index.html:4]:
```suggestion
<script src="//cdnjs.cloudflare.com/ajax/libs/socket.io/3.0.5/socket.io.js"></script>
```

File: rmf_demo_panel/task_list/airport_tasks.json [hunk 9]
```diff
@@ -0,0 +1 @@
+[{"task_type":"Delivery", "start_time":0, "description": {"option": "mop"}}]
```
Comment (Yadunund) [inline rmf_demo_panel/task_list/airport_tasks.json:1]: Let's delete the `rmf_demo_panel/task_list/` folder. The contents have already been moved here https://github.com/osrf/rmf_demos/tree/develop/dispatcher_demo/rmf_demo_tasks/rmf_demo_tasks

File: rmf_demo_assets/models/DeliveryRobot/model.sdf [hunk 10]
```diff
@@ -377,4 +377,3 @@
 
   </model>
 </sdf>
-
```
Comment (Yadunund) [inline rmf_demo_assets/models/DeliveryRobot/model.sdf:380]: Undo deletion

**Review by Yadunund — CHANGES_REQUESTED** [review]
Some minor changes to consider.

My main feedback is regarding the dashboard configuration files (dashboard_config.json). I was under the assumption that we were reusing (with changes) the files from [`rmf_dashboard_resources` ](https://github.com/osrf/rmf_demos/tree/master/rmf_dashboard_resources). But it looks like we are maintaining duplicates in this package. I'd recommend trying to reuse the files in `rmf_dashbaord_resources` so that we only have one set of configuration files to maintain. Perhaps we can look at how `rmf-web` retrieves the information from `rmf_dashboard_resources`. 

Lastly, for clarity purposes, could we change the title "Requests" and "Scheduled Task List" in the GUI to "Submit a task" and "Submit a list of tasks" respectively?

**Comment by Yadunund** [comment]
image


When a `Clean` task is selected, I think the label should read "Schedule a Clean Request" instead of Ad-hoc task.

File: rmf_demo_panel/rmf_demo_panel/static/dashboard_config.json (hunk 7 above)
Comment (vallq) [reply rmf_demo_panel/rmf_demo_panel/static/dashboard_config.json:1]: following from ^, I assumed that each world would have its own config file, so how it's implemented now is that clicking between tabs loads the different form options for the frontend. I think the outcome is that eventually the `dashboard_config` file would be appended to the `main.json` of each world in `rmf_dashboard_resources`

File: demos/launch/dispatcher.launch.xml (hunk 3 above)
Comment (vallq) [reply demos/launch/dispatcher.launch.xml:1]: hi Yadu, I think I misinterpreted our discussion in the office about the configuration json file previously, but I've clarified with @tanyouliang95 this afternoon

how rmf-web does it is through this [script ](https://github.com/osrf/rmf-web/blob/main/packages/dashboard/scripts/setup/setup.js) that the user has to run with `npm run setup` in the terminal before being able to launch the dashboard successfully. it provides a path to the resources needed for the dashboard to work (currently it allows retrieving the resources from a remote github repo or using an absolute path set by the user to retrieve the resources by creating a `.resources.json` file
image
resulting `.resources.json` file
`{"repoUrl":"https:github.com/osrf/rmf_demos.git","folder":"rmf_dashboard_resources/office/","branch":"master"}`

the current implementation for the rmf  panel assumes that that config file is available locally in the `static` folder, hence when the user navigates between the tabs, the form options are updated accordingly. 

I'm not sure if there can be relative paths using the way that rmf-web does it, I could look more into it, but in that case, do I assume that we want the config to be retrieved locally? instead of being hosted on a server and then the config is fetched with a HTTP request or smth like that?

On your last point, I've made the changes to the titles :+1:

**Comment by Yadunund** [comment]
@vallq thanks for addressing all the issues and for the explanation on how `rmf-web` manages the config files. 

Just thinking out loud here: what do you think of storing the config files in the local `rmf_dashboard_resources` pkg and then copying them into the `static` folder when we run `npm run setup` ?

File: demos/launch/dispatcher.launch.xml (hunk 3 above)
Comment (tanyouliang95) [reply demos/launch/dispatcher.launch.xml]: https://github.com/osrf/rmf_demos/pull/178. `common.launch.xml` will include `dashboard.launch.xml` and `dispactcher_node`.

**Comment by tanyouliang95** [comment]
@Yadunund  As proposed, `rmf_demo_panel` configs are now located in `rmf_dashboard_resources`. https://github.com/osrf/rmf_demos/pull/178

**Review by Yadunund — APPROVED** [review]
Works great! Thanks for making the necessary changes!

**Comment by vallq** [comment]
@tanyouliang95 fixed the bug in the `scheduled-tasks-form` that did not convert delivery requests with the right description

**Review by tanyouliang95 — APPROVED** [review]
Thanks! Tested and ready to merge!
